### DISCUSSION TITLE
Biometric Authentication with Uno

### DISCUSSION THREAD

**Discussion by trj2059** [unoplatform/uno#10914] [General]
Is there a way to use biometric authentication on the iOS and Android platforms with Uno?

**Comment by jeromelaban** [comment]
Thanks for the question! Uno Platform currently does not provide a WinRT API implementation related biometric authentication, but you can use the [native APIs directly](https://learn.microsoft.com/en-us/xamarin/android/platform/fingerprint-authentication/), and use [platform specific code](https://platform.uno/docs/articles/platform-specific-csharp.html) to handle the differences.

**Comment by jeanplevesque** [comment]
This open-source package offers an abstraction layer (and native implementations) for biometrics on iOS and Android for Uno apps.
https://github.com/nventive/BiometryService

**Reply by Mike-E-angelo** [reply]
Thank you for providing this link. 🙏  Do you have guidance on how this would be used for an Open ID Connect authentication? 🤔  I see there is an `Encrypt` and `Decrypt` function but am unclear how this would be used on a successful authentication.  Open ID Connect returns refresh tokens and those are already stored in `ITokenCache`.  Any suggestions would be appreciated.

**Reply by jeanplevesque** [reply]
We've typically used the BiometryService package to encrypt email-password credentials locally upon logging in successfully. I suppose you could do the same if the credentials are captured within the app (and not via a browser or external SDK).

**Reply by Mike-E-angelo** [reply]
OK thank you for that, that is useful.  In this case, the username (email address) + password is not in the process of the Uno application but within the browser application during the Open ID Connect authentication workflow.  

I suppose what we could do is attach a claim with a unique/generated value/token and return this on successful authentication.  Then back in the Uno app, store this in the Biometric service.  Upon successful biometric authentication, we could decrypt that token from the Biometric service and send it to an API controller on our Open ID identity server, validate the token somehow, and get a new authentication token for the user. 🤔  That's sort of what I am thinking about how to handle this at the moment and we'll see how I do. 😃🤞

**Reply by Mike-E-angelo** [reply]
I'm sort of slapping my forehead here, as this was way simpler than thought.  The out-of-the-box authentication from the Uno scaffolding template refreshes the Oauth2 token upon loading the first view via the `IAuthenticationService.RefreshAsync` call.  I ended up looking for that call to return `true` and if so, do a biometrics check on that to essentially verify the refresh.  So the lesson/point here is that biometrics doesn't necessarily have to save (encrypt/decrypt) something.  It can verify an existing process, too. 👍
